pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasurementsKokkos.hpp

```diff
@@ -509,117 +513,68 @@ class Measurements final
-                const PrecisionT REAL = arr_data(index).real();
-                const PrecisionT IMAG = arr_data(index).imag();
-                const PrecisionT value = REAL * REAL + IMAG * IMAG;
-                Kokkos::atomic_add(&d_probabilities(i), value);
-            });
-
-        std::vector<PrecisionT> probabilities(all_indices.size(), 0);
-
-        if (is_sorted_wires) {
-            Kokkos::deep_copy(UnmanagedPrecisionHostView(probabilities.data(),
-                                                         probabilities.size()),
-                              d_probabilities);
-            return probabilities;
+        Kokkos::View<ComplexT *> sv = this->_statevector.getView();
+
+        // Reducing over `d_probabilities` requires too much L0 scratch memory
+        // on GPUs. If n_wires >= 20, this also requires quite a bit of memory
+        // on CPUs, so we fallback to the next implementation
+        if (n_wires < 20 && !is_gpu_scratch_limited) {
+            using MDPolicyType_2D =
+                Kokkos::MDRangePolicy<Kokkos::Rank<2, Kokkos::Iterate::Left>>;
+            auto md_policy = MDPolicyType_2D(
+                {{0, 0}}, {{static_cast<int64_t>(all_indices.size()),
+                            static_cast<int64_t>(all_offsets.size())}});
+            Kokkos::parallel_reduce(
+                md_policy,
+                getProbsFunctor<PrecisionT, KokkosExecSpace>(
+                    sv, wires, d_all_indices, d_all_offsets),
+                d_probabilities);
         } else {
-            Kokkos::View<PrecisionT *> transposed_tensor("transposed_tensor",
-                                                         all_indices.size());
-
-            Kokkos::View<std::size_t *> d_trans_index("d_trans_index",
-                                                      all_indices.size());
-
-            const int num_trans_tensor = transposed_tensor.size();
-            const int num_sorted_ind_wires = sorted_ind_wires.size();
-
-            MDPolicyType_2D mdpolicy_2d1(
-                {{0, 0}}, {{num_trans_tensor, num_sorted_ind_wires}});
-
-            Kokkos::parallel_for(
-                "TransIndex", mdpolicy_2d1,
-                getTransposedIndexFunctor(d_sorted_ind_wires, d_trans_index,
-                                          num_sorted_ind_wires));
-
             Kokkos::parallel_for(
-                "Transpose",
-                Kokkos::RangePolicy<KokkosExecSpace>(0, num_trans_tensor),
-                getTransposedFunctor<PrecisionT>(
-                    transposed_tensor, d_probabilities, d_trans_index));
-
-            Kokkos::deep_copy(UnmanagedPrecisionHostView(probabilities.data(),
-                                                         probabilities.size()),
-                              transposed_tensor);
-
-            return probabilities;
+                all_indices.size(), KOKKOS_LAMBDA(const std::size_t i) {
+                    for (std::size_t j = 0; j < d_all_offsets.size(); j++) {
```

Comment:
Can we also parallelize the inner for-loop here?

Reply:
No, that basically the MDRange algo above which is slower in this regime.